File: java/euler/src/main/java/com/rbleggi/euler/P03.java
package com.rbleggi.euler;

public class P03 {
    static long largestPrimeFactor(long n) {
        long num = n;
        long factor = 2;
        long lastFactor = 1;
        while (factor * factor <= num) {
            if (num % factor == 0) {
                lastFactor = factor;
                num /= factor;
                while (num % factor == 0) {
                    num /= factor;
                }
            }
            factor += 1;
        }
        return num > 1 ? num : lastFactor;
    }

    public static void main(String[] args) {
        long testNumber = 13195L;
        long testResult = largestPrimeFactor(testNumber);
        System.out.println("Largest prime factor of " + testNumber + ": " + testResult);

        long targetNumber = 600851475143L;
        long result = largestPrimeFactor(targetNumber);
        System.out.println("Largest prime factor of " + targetNumber + ": " + result);
    }
}

File: kotlin/euler/src/main/kotlin/com/rbleggi/euler/P03.kt
package com.rbleggi.euler

fun mainP03() {
    fun largestPrimeFactor(n: Long): Long {
        var num = n
        var factor = 2L
        var lastFactor = 1L
        while (factor * factor <= num) {
            if (num % factor == 0L) {
                lastFactor = factor
                num /= factor
                while (num % factor == 0L) {
                    num /= factor
                }
            }
            factor += 1
        }
        return if (num > 1) num else lastFactor
    }

    val testNumber = 13195L
    val testResult = largestPrimeFactor(testNumber)
    println("Largest prime factor of $testNumber: $testResult")

    val targetNumber = 600851475143L
    val result = largestPrimeFactor(targetNumber)
    println("Largest prime factor of $targetNumber: $result")
}
